### PR TITLE
Prevent registering synced data maps for non-synced datapack registries

### PR DESCRIPTION
**Pull request by Matyrobbrt**
The sync will never work and will only lead to errors since the registry won't exist on the client.

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/registries/DataPackRegistriesHooks.java [hunk 1]
```diff
@@ -70,4 +71,11 @@ public static Stream<RegistryDataLoader.RegistryData<?>> getDataPackRegistriesWi
     public static Set<ResourceKey<? extends Registry<?>>> getSyncedCustomRegistries() {
         return SYNCED_CUSTOM_REGISTRIES_VIEW;
     }
+
+    @Nullable
+    @ApiStatus.Internal
+    @SuppressWarnings("unchecked")
+    public static <T> RegistrySynchronization.NetworkedRegistryData<T> getSyncedRegistry(final ResourceKey<? extends Registry<T>> registry) {
```
Comment (dhyces) [inline src/main/java/net/neoforged/neoforge/registries/DataPackRegistriesHooks.java:78]: why not a `isRegistrySynced` method instead that checks the keySet?

File: src/main/java/net/neoforged/neoforge/registries/datamaps/RegisterDataMapTypesEvent.java [hunk 2]
```diff
@@ -31,10 +32,17 @@ public RegisterDataMapTypesEvent(Map<ResourceKey<Registry<?>>, Map<ResourceLocat
      * @param type the data map type to register
      * @param <T>  the type of the data map
      * @param <R>  the type of the registry
-     * @throws IllegalArgumentException if a type with the same ID has already been registered for that registry
+     * @throws IllegalArgumentException      if a type with the same ID has already been registered for that registry
+     * @throws UnsupportedOperationException if the registry is a non-synced datapack registry and the data map is synced
      */
     public <T, R> void register(DataMapType<R, T> type) {
         final var registry = type.registryKey();
+        if (DataPackRegistriesHooks.getDataPackRegistries().stream().anyMatch(data -> data.key().equals(registry))) {
```
Comment (embeddedt) [inline src/main/java/net/neoforged/neoforge/registries/datamaps/RegisterDataMapTypesEvent.java:40]: Not a huge fan of needing to iterate over every registry here, is it not possible to expose a `contains` method of some sort?

Comment (dhyces) [reply src/main/java/net/neoforged/neoforge/registries/datamaps/RegisterDataMapTypesEvent.java:40]: I deleted my old comment based on this, but honestly it's unlikely for there to be so many DP regs for this to really matter. Aside from performance however, is it not possible to simply use `List#contains`?

Comment (Matyrobbrt) [reply src/main/java/net/neoforged/neoforge/registries/datamaps/RegisterDataMapTypesEvent.java:40]: The list isn't a `List<ResourceKey>`, it's a list of a record that also holds the codec, contains wouldn't work. I don't think that it's going to be an issue considering that there aren't thousands of dp regs.

**Review by marchermans — APPROVED** [review]

**Pull request merged**